src/parquet/column/reader.h

```diff
@@ -47,6 +47,66 @@ struct hash<parquet::Encoding::type> {
 
 namespace parquet_cpp {
 
+class LevelDecoder {
+
+ private:
+  parquet::Encoding::type encoding_;
+  std::unique_ptr<RleDecoder> rle_decoder;
+  std::unique_ptr<BitReader> bit_packed_decoder;
+
+ public:
+  LevelDecoder() {}
+
+  //return true if one of the decoder is set
+  bool set() {
+     return ((NULL != rle_decoder.get()) || (NULL != bit_packed_decoder.get()));
+  }
+
+  // Initialize the LevelDecoder.
+  void Init(parquet::Encoding::type encoding, int16_t max_level, int num_buffered_values,
+             const uint8_t** data, int* data_size) {
+
+     uint32_t num_bytes = 0;
+     encoding_ = encoding;
+     switch (encoding) {
+       case parquet::Encoding::RLE: {
+         num_bytes = *reinterpret_cast<const uint32_t*>(*data);
+         *data += sizeof(uint32_t);
+         *data_size -= sizeof(uint32_t);
+         int bit_width = BitUtil::Log2(max_level + 1);
+         rle_decoder.reset(new RleDecoder(*data, num_bytes, bit_width));
+         break;
+       }
+       case parquet::Encoding::BIT_PACKED: {
+         num_bytes = BitUtil::Ceil(num_buffered_values, 8);
+         bit_packed_decoder.reset(new BitReader(*data, num_bytes));
+         break;
+       }
+       default:
+         throw ParquetException("Unknown encoding type for levels.");
+     }
+     *data += num_bytes;
+     *data_size += num_bytes;
+  }
+
+  /// Returns the next level.
+  inline int16_t ReadLevel(int16_t *level){
```

Comment:
Per discussion above, the public API for this class should be array oriented (with `batch_size` as a parameter) and return the actual number of levels read (in case the decoder reaches EOS before reaching the batch size -- will need some tests around this)